WorkflowSwiftUI/Sources/ObservableScreen.swift

```diff
@@ -146,21 +147,31 @@ private final class ModeledHostingController<Model, Content: View>: UIHostingCon
 
         defer { hasLaidOutOnce = true }
 
-        if #available(iOS 16.0, *) {
-            // Handled in initializer, but set it on first layout to resolve a bug where the PCS is
-            // not updated appropriately after the first layout.
-            // UI-5797
-            if !hasLaidOutOnce,
-               swiftUIScreenSizingOptions.contains(.preferredContentSize)
-            {
-                let size = view.sizeThatFits(view.frame.size)
-
-                if preferredContentSize != size {
-                    preferredContentSize = size
-                }
-            }
-        } else if swiftUIScreenSizingOptions.contains(.preferredContentSize) {
-            let size = view.sizeThatFits(view.frame.size)
+        if swiftUIScreenSizingOptions.contains(.preferredContentSize) {
+            // Use the largest frame ever laid out in as a constraint for preferredContentSize
+            // measurements.
+            let width = max(view.frame.width, maxFrameWidth)
+            let height = max(view.frame.height, maxFrameHeight)
+
+            maxFrameWidth = width
+            maxFrameHeight = height
+
+            let fixedSize = CGSize(width: width, height: height)
+
+            // Measure a few different ways to account for ScrollView behavior. ScrollViews will
+            // always greedily fill the space available, but will report the natural content size
+            // when given an infinite size. By combining the results of these measurements we can
+            // deduce the natural size of content that scrolls in either direction, or both, or
+            // neither.
+
+            let fixedResult = view.sizeThatFits(fixedSize)
+            let unboundedHorizontalResult = view.sizeThatFits(CGSize(width: .infinity, height: fixedSize.height))
+            let unboundedVerticalResult = view.sizeThatFits(CGSize(width: fixedSize.width, height: .infinity))
+
+            let size = CGSize(
+                width: min(fixedResult.width, unboundedHorizontalResult.width),
+                height: min(fixedResult.height, unboundedVerticalResult.height)
+            )
```

Comment:
any chance we need to (or should) handle weird/unexpected values here? is `sizeThatFits` guaranteed to return sensical results? i guess the concern would be if we end up assigning a preferred content size that doesn't 'make sense' – would that matter?

Reply:
> any chance we need to (or should) handle weird/unexpected values here? is `sizeThatFits` guaranteed to return sensical results? i guess the concern would be if we end up assigning a preferred content size that doesn't 'make sense' – would that matter?

It's not guaranteed to do anything by API contract, but based on observation it seems like SwiftUI will return sensible values.